### PR TITLE
Fix pickling of AxesWidgets.

### LINKED ISSUE #22482: [ENH]:  pickle (or save) matplotlib figure with insteractive slider
### Problem

I'd like to save/pickle a matplotlib figure containing a `matplotlib.widgets.Widget` object to send it to a colleague without him having to run the code I use to generate it.

I know since matplotlib version 1.2 `matplotlib.figure.Figure` are pickable. However, when the figure contains a `matplotlib.widgets.Widget`, pickle fails to save it.

For example consider this class:

```
import matplotlib.pyplot as plt
import matplotlib.widgets as widgets

import numpy as np

import pickle as pkl

class interactive_plot():

    def __init__(self, add_slider):

        self.fig, self.axs = plt.subplots(2,1, gridspec_kw={'height_ratios':[10,1]})

        self.x_plot = np.linspace(0,10,1000)
        self.line = self.axs[0].plot(self.x_plot, self.x_plot*0)[0]
        self.axs[0].set_ylim((0,10))
        self.axs[0].grid()
        self.axs[0].set_title(r'$y = x^\alpha$')

        if add_slider:
            self.slider= widgets.Slider(
                self.axs[1],
                label   = r'$\alpha$',
                valmin  = 0,
                valmax  = 5,
                valinit = 0,
                valstep = .01
                )
            self.slider.on_changed(self.update)

    def update(self, val):
        self.line.set_ydata(self.x_plot**val)
```

If I run

```
h = interactive_plot(add_slider=False)
with open('test.pkl','wb') as f: pkl.dump(h, f)
```
all works properly. However
```
h = interactive_plot(add_slider=True)
with open('test.pkl','wb') as f: pkl.dump(h, f)
```
returns
```
Traceback (most recent call last):

  File "C:\Users\user\Desktop\test.py", line 47, in <module>
    with open('test.pkl','wb') as f: pkl.dump(h, f)

TypeError: cannot pickle 'FigureCanvasQTAgg' object
```

Is there a way to workaround this?

DISCLAIMER: The same question was posted here https://stackoverflow.com/questions/71145817/pickle-or-save-matplotlib-figure-with-insteractive-slider

### Proposed solution

_No response_

### PR DESCRIPTION
**Pull request by anntzer**
There's no need to hold onto the (non-picklable) canvas as an attribute.

Closes #22482.

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by QuLogic — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Review by QuLogic — CHANGES_REQUESTED** [review]
This apparently breaks `SpanSelector`.

**Comment by tacaswell** [comment]
This is also going to break `_changed_canvas` which we use to suppress events durring saving (see https://github.com/matplotlib/matplotlib/pull/25085 / 6bb208d5a9c46ded307c7baef4fd255cdbc80c46 )

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -2588,14 +2571,12 @@ def __init__(self, ax, onselect, direction, *, minspan=0, useblit=False,
 
         self._active_handle = None
 
-    def new_axes(self, ax, *, _props=None):
+    def new_axes(self, ax, *, _props=None, _init=False):
         """Set SpanSelector to operate on a new Axes."""
         self.ax = ax
-        if self.canvas is not ax.figure.canvas:
+        if _init or self.canvas is not ax.figure.canvas:
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py:2577]:
```suggestion
        if _init:
```

The second part will never be true now.

**Comment by anntzer** [comment]
Now fixed; the test failure is unrelated.

**Comment by ksunden** [comment]
Test failure is not unrelated, it is the newly added test

File: lib/matplotlib/widgets.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/widgets.py:2577]: It was pointed out I am wrong about being wrong.

**Review by QuLogic — APPROVED** [review]

**Review by ksunden — APPROVED** [review]
Anyone can merge on CI finishing

**Pull request merged**